cmd/osctl/cmd/top.go

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package cmd
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"time"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/gizak/termui/v3/widgets"
+	"github.com/spf13/cobra"
+	"github.com/talos-systems/talos/cmd/osctl/pkg/client"
+	"github.com/talos-systems/talos/cmd/osctl/pkg/helpers"
+	"github.com/talos-systems/talos/internal/pkg/constants"
+	"github.com/talos-systems/talos/internal/pkg/proc"
+)
+
+// versionCmd represents the version command
+var topCmd = &cobra.Command{
+	Use:   "top",
+	Short: "Streams top output",
+	Long:  ``,
+	Run: func(cmd *cobra.Command, args []string) {
+		creds, err := client.NewDefaultClientCredentials(talosconfig)
+		if err != nil {
+			helpers.Fatalf("error getting client credentials: %s", err)
+		}
+		c, err := client.NewClient(constants.OsdPort, creds)
+		if err != nil {
+			helpers.Fatalf("error constructing client: %s", err)
+		}
+
+		if err := ui.Init(); err != nil {
+			log.Fatalf("failed to initialize termui: %v", err)
+		}
+		defer ui.Close()
+
+		topUI(c)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(topCmd)
+}
+
+func topUI(c *client.Client) {
+
+	l := widgets.NewList()
+	l.Title = "Top"
+	// TODO see if we can dynamically get this
+	// x, y, w, h
+	l.SetRect(0, 0, 65, 30)
+	l.TextStyle.Fg = ui.ColorYellow
+
+	draw := func(procs []proc.ProcessList) {
+		rss := func(p1, p2 *proc.ProcessList) bool {
+			// Reverse sort ( Descending )
+			return p1.ResidentMemory > p2.ResidentMemory
+		}
+		By(rss).Sort(procs)
+		s := make([]string, 0, len(procs))
+		s = append(s, fmt.Sprintf("%s %s %s %s %s %s %s %s", "PID", "State", "Threads", "CPU Time", "VirtMem", "ResMem", "Command", "Exec/Args"))
+		for _, p := range procs {
+			//log.Printf("Sorted RSS: %+v", p)
+			s = append(s, p.String())
+		}
+		l.Rows = s
+
+		ui.Render(l)
+	}
+
+	procs, err := c.Top()
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	draw(procs)
+	uiEvents := ui.PollEvents()
+	ticker := time.NewTicker(time.Second).C
+	for {
+		select {
+		case e := <-uiEvents:
+			switch e.ID {
+			case "q", "<C-c>":
+				return
+			}
+		case <-ticker:
+			procs, err := c.Top()
+			if err != nil {
+				log.Println(err)
+				return
+			}
+			draw(procs)
+		}
+	}
+}
+
+type By func(p1, p2 *proc.ProcessList) bool
```

Comment:
This chunk is largely based on https://golang.org/pkg/sort/ `Example (SortKeys)`